x-pack/plugin/profiler/src/main/java/org/elasticsearch/xpack/profiler/Migration.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.profiler;
+
+import org.elasticsearch.action.admin.indices.mapping.put.PutMappingRequest;
+import org.elasticsearch.action.admin.indices.settings.put.UpdateSettingsRequest;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.function.Consumer;
+
+/**
+ * Encapsulates a single migration action.
+ */
+public interface Migration {
+    /**
+     * @return The index template version that the targeted index will be on after the migration has been applied.
+     */
+    int getTargetIndexTemplateVersion();
+
+    /**
+     * Applies this migration. Depending on the type of migration one or the other consumer will be called.
+     *
+     * @param index The name of the specific index to which the migration should be applied.
+     * @param mappingConsumer A consumer that applies mapping changes. Must not be null.
```

Comment:
Why not annotating the non-null arguments with `@NonNull` ?

Reply:
The established practice in the code base is to annotate nullable parameters with `@Nullable` instead so I could probably just remove that bit in the Javadoc comment.

Reply:
Done in https://github.com/elastic/elasticsearch/pull/97773/commits/8227c5bc26586f8678aec4cb0b55d1c34fc8378f